HeterogeneousCore/CUDAUtilities/interface/maxCoopBlocks.h

```diff
@@ -0,0 +1,30 @@
+
+#ifndef HeterogeneousCore_CUDAUtilities_interface_maxCoopBlocks_h
+#define HeterogeneousCore_CUDAUtilities_interface_maxCoopBlocks_h
+
+#include <cuda_runtime.h>
+#include <algorithm>
+#include <cstdlib>
+
+template <typename F>
+inline int maxCoopBlocks(F kernel, int nthreads, int shmem, int device, int redFact = 10) {
+#define GET_COOP_RED_FACT_FROM_ENV
+
+// to drive performance assessment by envvar
+#ifdef GET_COOP_RED_FACT_FROM_ENV
+  auto env = getenv("COOP_RED_FACT");
+  int redFactFromEnv = env ? atoi(env) : 0;
+  if (redFactFromEnv != 0)
+    redFact = redFactFromEnv;
+#endif
+
+  int numBlocksPerSm = 0;
+  cudaDeviceProp deviceProp;
+  cudaGetDeviceProperties(&deviceProp, device);
```

Comment:
as I commented above I hope all this could be (at some point) encapsulated in a service

Reply:
Right, but currently it seems that all calls to `maxCoopBlocks()` explicitly give `device = 0` argument. I was thinking that it would be better to either have the callers to pass the actual device ID, or this function to use `cms::cuda::currentDevice()` to figure it out.

On the other hand, as long as the "caching" is done with function-scope `static`s, in presence of multiple devices the function would be evaluated only one of them, so a mixed-device setup wouldn't really work anyway. So the current setup is as good as the alternative until a better way to cache the results.